### PR TITLE
Phase2-hgx151 Modify RecHitTools for servicing old and new geometries

### PR DESCRIPTION
**Pull request by bsunanda**
Doesn't need making any fake DetId and also wafer() and cell() now returns a ptr of integers (for old geometry the second int is 0)

### CODE REVIEW

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by kpedro88** [comment]
+upgrade
a few changes in 2023 workflows occur (miniAOD, slimmedElectronsFromMultiCl userFloats), as expected because of bug fixes in RecHitTools

**Comment by kpedro88** [comment]
please test workflow 24034.0

File: RecoLocalCalo/HGCalRecAlgos/src/RecHitTools.cc [hunk 1]
```diff
@@ -125,22 +142,11 @@ GlobalPoint RecHitTools::getPosition(const DetId& id) const {
...
-    if (layer <= fhOffset_) id = HGCSiliconDetId(DetId::HGCalEE, 1, 0, layer, 0, 0, 0, 0);
-    else if (layer <= bhOffset_) id = HGCSiliconDetId(DetId::HGCalHSi, 1, 0, layer - fhOffset_, 0, 0, 0, 0);
-    else  id = HGCScintillatorDetId(0, layer - bhOffset_, 1, 0);
-  }
-  return getPosition(id);
+GlobalPoint RecHitTools::getPositionLayer(int layer) const {
+  int lay = std::abs(layer);
+  const HGCalDDDConstants* ddd = get_ddd(geom_, geometryType_, fhOffset_, lay);
+  double z = (layer > 0) ? ddd->waferZ(lay,true) : -ddd->waferZ(lay,true);
+  return GlobalPoint(0,0,z);
```
[the remark this answers is not in the document]
Comment (slava77) [reply RecoLocalCalo/HGCalRecAlgos/src/RecHitTools.cc]: this seems OK, at least in 10_3_0_pre4, where the only use case is in RecoEgamma/EgammaTools/src/EgammaPCAHelper.cc
```return recHitTools_->getPositionLayer(firstLayer).z();```
Still, if only z is ever needed, why bother making this return a 3D point.

**Comment by slava77** [comment]
> Reco comparison results: 12 differences found in the comparisons

there are somewhat large changes in FromMultiCl electron and photon userFloats.
Are these expected?

e.g. wf 20034
all_mini_oldvsnew_ttbar14tev2023d17wf20034p0c_patelectrons_slimmedelectronsfrommulticl__reco_obj___userfloats__20_
all_mini_oldvsnew_ttbar14tev2023d17wf20034p0c_patelectrons_slimmedelectronsfrommulticl__reco_obj___userfloats__5_

**Comment by slava77** [comment]
if I extracted the names of the floats correctly, the changes are in depthCompatibility and measuredDepth variables of the photons and electrons.

File: RecoLocalCalo/HGCalRecAlgos/src/RecHitTools.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (bsunanda) [reply RecoLocalCalo/HGCalRecAlgos/src/RecHitTools.cc]: Well with layer information one cannot get x,y of a cell. It was artificially made into a global point. I can make it to getZPositionLayer to make it more explicit provided Marco etc agree to this.

**Comment by slava77** [comment]
@bsunanda @rovere @felicepantaleo 
please comment on the differences introduced by this PR in the old geometry workflows.

**Comment by kpedro88** [comment]
@clelange had reported that `getPositionLayer()` was returning incorrect z positions in some cases. There was also a small fix in `getSiThickIndex()`. These together can introduce the differences observed.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/24504	0c87527
- code changes are in line with the PR description. Based on https://github.com/cms-sw/cmssw/pull/24504#issuecomment-422796601 some changes (observed in the comparisons) are expected in the depth variable(s)
- jenkins tests pass and comparisons with the baseline show differences only in phase-2 HGCAL high level electron and photon variables related to depth/layer-position .


The ```return GlobalPoint(0,0,z).z() -> return z``` conversion discussed in https://github.com/cms-sw/cmssw/pull/24504#pullrequestreview-155609729 can be done separately.

**Comment by fabiocos** [comment]
+1

**Pull request merged**